datacompy/fuguecompare.py

```diff
@@ -0,0 +1,189 @@
+        Boolean indicator that controls of column names will be cast into lower case
+    """
+    if (
+        isinstance(df1, pd.DataFrame)
+        and isinstance(df2, pd.DataFrame)
+        and parallelism is None  # user did not specify parallelism
+        and fa.get_current_parallelism() == 1  # currently on a local execution engine
+    ):
+        comp = Compare(
+            df1=df1,
+            df2=df2,
+            join_columns=join_columns,
+            abs_tol=abs_tol,
+            rel_tol=rel_tol,
+            df1_name=df1_name,
+            df2_name=df2_name,
+            ignore_spaces=ignore_spaces,
+            ignore_case=ignore_case,
+            cast_column_names_lower=cast_column_names_lower,
+        )
+        return comp.matches()
+
+    tdf1 = fa.as_fugue_df(df1)
+    tdf2 = fa.as_fugue_df(df2)
+
+    if isinstance(join_columns, str):
+        hash_cols = [join_columns]
+    else:
+        hash_cols = join_columns
+    # ensure all cols exist
+    hash_cols = tdf1.schema.extract(hash_cols).names
+
+    if cast_column_names_lower:
+        tdf1 = tdf1.rename(
+            {col: col.lower() for col in tdf1.schema.names if col != col.lower()}
+        )
+        tdf2 = tdf2.rename(
+            {col: col.lower() for col in tdf2.schema.names if col != col.lower()}
+        )
+        hash_cols = [col.lower() for col in hash_cols]
+
+    if tdf1.schema != tdf2.schema:
+        return False
+
+    all_cols = tdf1.schema.names
+    str_cols = set(f.name for f in tdf1.schema.fields if pa.types.is_string(f.type))
+    bucket = (
+        parallelism if parallelism is not None else fa.get_current_parallelism() * 2
+    )
+
+    def _serialize(dfs: Iterable[pd.DataFrame], left: bool) -> Iterable[Dict[str, Any]]:
+        for df in dfs:
+            cols = {}
+            for name in df.columns:
+                col = df[name]
+                if name in str_cols:
+                    if ignore_spaces:
+                        col = col.str.strip()
+                    if ignore_case:
+                        col = col.str.lower()
+                cols[name] = col
+            data = pd.DataFrame(cols)
+            gp = pd.util.hash_pandas_object(df[hash_cols], index=False).mod(bucket)
+            for k, sub in data.groupby(gp, as_index=False, group_keys=False):
+                yield {"key": k, "left": left, "data": pickle.dumps(sub)}
+
+    ser = fa.union(
+        fa.transform(
+            tdf1,
+            _serialize,
+            schema="key:int,left:bool,data:binary",
+            params=dict(left=True),
+        ),
+        fa.transform(
+            tdf2,
+            _serialize,
+            schema="key:int,left:bool,data:binary",
+            params=dict(left=False),
+        ),
+        distinct=False,
+    )
+
+    def _comp(df: List[Dict[str, Any]]) -> List[List[Any]]:
+        df1 = (
+            pd.concat([pickle.loads(r["data"]) for r in df if r["left"]])
+            .sort_values(all_cols)
```

Comment:
Similar question as above. `all_cols`, `join_columns`, `abs_tol`, etc.. are all being picked up from outside of `_comp`. Might be OK but just thought I'd ask.

Reply:
Same, this is why I use the embedded function.